### ISSUE TITLE
DNS failures in pods on new and rebooted worker units

### ISSUE THREAD

**Issue by jetpackdanger** [juju-solutions/bundle-canonical-kubernetes#502]
We recently rebooted a couple of our kubernetes-worker units, one each in a pair of CDK environments.  Following the reboot, in both cases, DNS is no longer working in containers running on the rebooted worker.

There is some evidence that DNS works for a brief time, at least occassionally, in that our appserver pods successfully look up the kube-dns alias for the external database host and connect, but when I kubectl exec into the pod and try DNS lookups, DNS hangs and times out.

The deployment is running Kubernetes 1.8.8 on Ubuntu 16.04 LTS.

**Comment by axinojolais** [comment]
I just solved this problem by running `sudo iptables -P FORWARD ACCEPT` on all the worker nodes. ufw sets this to DENY by default (in `/etc/default/ufw`).

**Comment by ktsakalozos** [comment]
Thank you @jetpackdanger , @axinojolais , for reporting and offering a fix for this issue. Would you please provide some more info? What provider are we talking about here, openstack? have you seen this in other providers as well? Do you have any custom iptable rules you apply yourself on the workers?  Is the `sudo iptables -P FORWARD ACCEPT` something you would expect to be there on all CDK deployments or is it specific to your setup?

Thanks.

**Comment by axinojolais** [comment]
Yes, provider is openstack. We don't use any other provider, so can't report. This bug is probably present wherever ufw is installed. We don't have/need custom iptables rules.

I don't think setting the FORWARD policy to ACCEPT is specific to our setup. In fact, it's mentioned at https://docs.docker.com/install/linux/linux-postinstall/#allow-access-to-the-remote-api-through-a-firewall. I'm not sure if anyone is using ufw on juju-deployed instances though. Perhaps you'll want to be more restrictive than simply unconditionally allowing any forward.

**Comment by mthaddon** [comment]
I see this in the charm:

```
def fix_iptables_for_docker_1_13():
     """ Fix iptables FORWARD policy for Docker >=1.13
     https://github.com/kubernetes/kubernetes/issues/40182
     https://github.com/kubernetes/kubernetes/issues/39823
     """
     cmd = ['iptables', '-w', '300', '-P', 'FORWARD', 'ACCEPT']
     check_call(cmd)
```

Sounds like that needs to persist reboots.

**Comment by mthaddon** [comment]
Actually, it's possible we need to update our charms, we don't see that code in our charms on disk

**Comment by jetpackdanger** [comment]
ufw is not enabled on any of our kubernetes-worker units, so I don't believe this is the source of the bogus FORWARD chain policy.

**Comment by jetpackdanger** [comment]
I replaced /sbin/iptables with a wrapper script and it looks like docker-containerd is to blame for this.

ubuntu@juju-0f73fb-stg-is-kubernetes-77:~$ grep -B1 -- '-P FORWARD DROP' /tmp/iptables.*
/tmp/iptables.2155.ps.dump-root      1559  2.0  0.0 282524 12316 ?        Ssl  20:17   0:00  \_ containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/containerd --shim containerd-shim --runtime runc
/tmp/iptables.2155.ps.dump:root      2155  0.0  0.0   4504   840 ?        S    20:17   0:00  \_ /bin/sh /sbin/iptables --wait -t filter -P FORWARD DROP
--
/tmp/iptables.2157.ps.dump-root      1559  2.0  0.0 282524 12316 ?        Ssl  20:17   0:00  \_ containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/containerd --shim containerd-shim --runtime runc
/tmp/iptables.2157.ps.dump:root      2155  0.0  0.0   4504   840 ?        R    20:17   0:00  \_ /bin/sh /sbin/iptables --wait -t filter -P FORWARD DROP
ubuntu@juju-0f73fb-stg-is-kubernetes-77:~$

**Comment by jetpackdanger** [comment]
...which is in fact what the docstring in that code Tom pasted is saying. (I had read it as docker >= 1.13 _failing to cope_ with FORWARD being set to DROP, rather than being responsible for it.)

**Comment by hloeung** [comment]
https://github.com/moby/moby/pull/28257
https://github.com/kubernetes/kubernetes/issues/40182
